File: app/schemas/indexes.py
"""Schemas for the backend-aware index management API (`/api/indexes`)."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field, model_validator

from app.schemas.enums import IndexBackend


class IndexRead(BaseModel):
    """Serialized vector-index metadata, tagged with its backend."""

    name: str
    backend: IndexBackend
    vector_type: str | None = None
    metric: str | None = None
    dimension: int | None = None
    status: dict[str, Any] | None = None
    host: str | None = None
    spec: dict[str, Any] | None = None
    deletion_protection: str | None = None
    tags: dict[str, str] | None = None


class IndexList(BaseModel):
    """List response for vector indexes."""

    indexes: list[IndexRead] = Field(default_factory=list)


class IndexCreateRequest(BaseModel):
    """Payload for creating a vector index on a chosen backend.

    `cloud`/`region`/`deletion_protection`/`tags` and `vector_type="sparse"`
    are Pinecone-only; capability validation in `IndexAdminService.create`
    rejects unsupported combinations with the backend's declared limits.
    """

    backend: IndexBackend
    name: str = Field(min_length=1, max_length=45)
    vector_type: str = Field(default="dense")
    dimension: int | None = Field(default=None, gt=0)
    metric: str = Field(default="cosine")
    cloud: str | None = None
    region: str | None = None
    deletion_protection: str | None = None
    tags: dict[str, str] | None = None

    @model_validator(mode="after")
    def validate_dimension(self) -> IndexCreateRequest:
        """Ensure dense indexes define a dimension and sparse indexes do not."""
        if self.vector_type == "dense" and self.dimension is None:
            raise ValueError("Dense indexes require a dimension.")
        if self.vector_type == "sparse" and self.dimension is not None:
            raise ValueError("Sparse indexes must not define a dimension.")
        return self


class IndexDeleteResponse(BaseModel):
    """Response after deleting a vector index."""

    status: str = "deleted"


class BackendCapabilitiesRead(BaseModel):
    """A backend's limits, served so the UI clamps inputs from live data."""

    max_dimension: int
    supported_metrics: list[str]
    supported_vector_types: list[str]
    index_name_max_length: int
    max_upsert_batch: int
    max_top_k: int
    requires_api_key: bool


class BackendInfoRead(BaseModel):
    """One vector-store backend's usability for the current user."""

    backend: IndexBackend
    label: str
    available: bool
    configured: bool
    capabilities: BackendCapabilitiesRead


class BackendInfoList(BaseModel):
    """List response for `GET /api/indexes/backends`."""

    backends: list[BackendInfoRead] = Field(default_factory=list)

File: frontend/src/test/fixtures/indexes.ts
import type { BackendInfo, VectorIndex } from "@/lib/types";

export function makeVectorIndex(overrides: Partial<VectorIndex> = {}): VectorIndex {
  return {
    name: "index-1",
    backend: "pgvector",
    vector_type: "dense",
    metric: "cosine",
    dimension: 1536,
    status: { ready: true, state: "Ready" },
    host: "index-1.pinecone.io",
    deletion_protection: "disabled",
    ...overrides,
  };
}

export function makeBackendInfo(overrides: Partial<BackendInfo> = {}): BackendInfo {
  return {
    backend: "pgvector",
    label: "pgvector (PostgreSQL)",
    available: true,
    configured: true,
    lexical_available: true,
    capabilities: {
      max_dimension: 2000,
      supported_metrics: ["cosine", "l2", "dotproduct"],
      supported_vector_types: ["dense", "sparse"],
      index_name_max_length: 45,
      max_upsert_batch: 1000,
      max_top_k: 10000,
      requires_api_key: false,
      supports_lexical_count: true,
      supports_lexical_facet: true,
    },
    ...overrides,
  };
}

export function makePineconeBackendInfo(overrides: Partial<BackendInfo> = {}): BackendInfo {
  return makeBackendInfo({
    backend: "pinecone",
    label: "Pinecone",
    configured: true,
    capabilities: {
      max_dimension: 20000,
      supported_metrics: ["cosine", "euclidean", "dotproduct"],
      supported_vector_types: ["dense", "sparse"],
      index_name_max_length: 45,
      max_upsert_batch: 1000,
      max_top_k: 10000,
      requires_api_key: true,
      // Pinecone serves sparse indexes but has no count/facet aggregation API.
      supports_lexical_count: false,
      supports_lexical_facet: false,
    },
    ...overrides,
  });
}
